src/backend/src/user_profile_model.rs

```diff
@@ -0,0 +1,163 @@
+use crate::types::{Candid, StoredPrincipal, UserProfileMap, UserProfileUpdatedMap};
+use shared::types::{user_profile::StoredUserProfile, Timestamp};
+
+pub struct UserProfileModel<'a> {
+    user_profile_map: &'a mut UserProfileMap,
+    user_profile_updated_map: &'a mut UserProfileUpdatedMap,
+}
+
+/// `UserProfileModel` should be used to access and manage the state to user profiles in the stable memory
+impl<'a> UserProfileModel<'a> {
+    pub fn new(
+        user_profile_map: &'a mut UserProfileMap,
+        user_profile_updated_map: &'a mut UserProfileUpdatedMap,
+    ) -> UserProfileModel<'a> {
+        UserProfileModel {
+            user_profile_map,
+            user_profile_updated_map,
+        }
+    }
+
+    pub fn find_by_principal(&self, user_principal: StoredPrincipal) -> Option<StoredUserProfile> {
+        if let Some(updated) = self.user_profile_updated_map.get(&user_principal) {
+            self.user_profile_map
+                .get(&(updated, user_principal))
+                .map(|p| p.0)
+        } else {
+            None
+        }
+    }
+
+    pub fn store_new(
+        &mut self,
+        user_principal: StoredPrincipal,
+        timestamp: Timestamp,
+        new_user: &StoredUserProfile,
+    ) {
+        if let Some(old_updated) = self.user_profile_updated_map.get(&user_principal) {
+            // Clean up old entries
+            self.user_profile_map.remove(&(old_updated, user_principal));
+        }
+        self.user_profile_updated_map
+            .insert(user_principal, timestamp);
+        self.user_profile_map
+            .insert((timestamp, user_principal), Candid(new_user.clone()));
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I believe that consistency of the `UserProfileModel` can be defined by:

* The number of entries in the two underlying maps is the same.
* For every entry in the principal->timestamps map, there is a corresponding entry in the other map.

I would recommend making a method like this, and checking consistency after any CRUD operations:

```
untested pseudocode!
impl UserProfileModel {
    #[cfg(test)]  // Compiled in only in test code.  Not needed in production.
    fn assert_consistent(&self) {
      assert_eq!(self.user_profile.len(), self.user_profile_timestamps.len(), "The number of entries should be the same");
      for (principal, timestamp) in self.... {
         assert!(self.user_profiles.contains((timestamp, principal)), "...");
      }
    }
}
```
This way it becomes possible to test integrity easily, just by calling `.assert_consistent()`, even after we add various other methods or in proptests.  What do you think?

Reply:
Done!